flink-connector-rabbitmq/src/main/java/org/apache/flink/connector/rabbitmq/sink/RabbitMQSinkWriter.java

```diff
@@ -0,0 +1,260 @@
+        }
+
+        inflightMessages.put(sequenceNumber, recordMessage);
+    }
+
+    private ConfirmCallback getSuccessConfirmCallback() {
+        return (seqNo, acknowledgePrevious) ->
+                mailboxExecutor.execute(
+                        () -> {
+                            if (acknowledgePrevious) {
+                                LOG.debug(
+                                        "Acknowledge all messages with sequence number less than or equal to {}",
+                                        seqNo);
+                                while (!inflightMessages.isEmpty()
+                                        && inflightMessages.firstKey() <= (Long) seqNo) {
+                                    RabbitMQMessage<T> message =
+                                            inflightMessages.remove(inflightMessages.firstKey());
+                                    numBytesOutCounter.inc(
+                                            serializationSchema.serialize(message.getMessage())
+                                                    .length);
+                                    numRecordsOutCounter.inc();
+                                }
+                            } else {
+                                LOG.debug("Acknowledge message with sequence number {}", seqNo);
+                                RabbitMQMessage<T> message = inflightMessages.remove(seqNo);
+                                numBytesOutCounter.inc(
+                                        serializationSchema.serialize(message.getMessage()).length);
+                                numRecordsOutCounter.inc();
+                            }
+                        },
+                        "Acknowledge message with sequence number " + seqNo);
+    }
+
+    private ConfirmCallback getFailureConfirmCallback() {
+        return (seqNo, acknowledgePrevious) ->
+                mailboxExecutor.execute(
+                        () -> {
+                            if (failOnError) {
+                                LOG.error(
+                                        "Failed to send message with sequence number {} and payload {}",
+                                        seqNo,
+                                        inflightMessages.get(seqNo).getMessage());
+                                throw new FlinkRuntimeException(
+                                        String.format(
+                                                "Failed to send message with sequence number %d and payload %s",
+                                                seqNo, inflightMessages.get(seqNo).getMessage()));
+                            }
+                            LOG.warn(
+                                    "Resending failed message with sequence number {} and payload {}",
+                                    seqNo,
+                                    inflightMessages.get(seqNo).getMessage());
+                            publishMessage(inflightMessages.remove(seqNo));
```

Comment:
Similarly we can think about retry threshold

Reply:
I agree, we can also address in a follow up

Reply:
Yeap let's address it in separate issue